include/nil/blueprint/lookup_table_library.hpp

```diff
@@ -0,0 +1,142 @@
+        }
+
+        template <typename Type, typename = void>
+        struct component_use_custom_lookup_tables : std::false_type{ };
+
+        template <typename Type>
+        struct component_use_custom_lookup_tables<Type, 
+            typename std::enable_if<std::is_member_function_pointer<decltype(&Type::component_custom_lookup_tables)>::value>::type> : std::true_type
+        { };
+
+        template <typename Type>
+        constexpr bool use_custom_lookup_tables(){
+            if(component_use_custom_lookup_tables<Type>::value){
+                return true;
+            }
+            return false;
+        }
+
+        template <typename BlueprintFieldType>
+        class lookup_table_library{
+            using lookup_table_definition = typename nil::crypto3::zk::snark::detail::lookup_table_definition<BlueprintFieldType>;
+            using filled_lookup_table_definition = typename nil::crypto3::zk::snark::detail::filled_lookup_table_definition<BlueprintFieldType>;
+
+            class binary_xor_table_type : public lookup_table_definition{
+            public:
+                binary_xor_table_type(): lookup_table_definition("binary_xor_table"){
+                    this->subtables["full"] = {{0,1,2}, 0, 3};
+                }
+                virtual void generate(){
+                    this->_table = {
+                        {0, 0, 1, 1},
+                        {0, 1, 0, 1},
+                        {0, 1, 1, 0}
+                    };
+                }
+            };
+        protected:
+            binary_xor_table_type binary_xor_table;
+
+            std::map<std::string, lookup_table_definition*> tables;
+            std::set<std::string> reserved_tables;
+            std::map<std::string, std::size_t> reserved_tables_indices;
+            std::map<std::string, lookup_table_definition*> reserved_tables_map;
+            // Last index
+        public:
+            lookup_table_library(){
+                tables = {};
+                tables["binary_xor_table"] = &binary_xor_table;
+            }
+
+            void register_lookup_table(const lookup_table_definition &table){
+                BOOST_ASSERT(tables.find(table.table_name) == tables.end());
+                tables[table.table_name] = &table;
+            }
+
+            void reserve_table(std::string name){
+                std::string table_name = name.substr(0, name.find("/"));
+                BOOST_ASSERT(tables.find(table_name) != tables.end());
+                std::string subtable_name = name.substr(name.find("/")+1, name.size());
+                BOOST_ASSERT(tables[table_name]->subtables.find(subtable_name) != tables[table_name]->subtables.end());
+                reserved_tables.insert(name);
+            }
+
+            const std::map<std::string, std::size_t> &get_reserved_indices(){
+                reserved_tables_indices = {};
+                std::size_t i = 1;
+                for (auto &name : reserved_tables){
```

Comment:
This is incorrect: adding a new item into the `reserved_tables` set might cause it to be added at any order. So this function is going to return different indices each time.
We should change the `reserved_tables_indices` only when adding a new table; we should also keep track of the last added index.

Reply:
get_reserved_indices call has sense only after all necessary tables are reserved.
std::set is ordered set.

So, this function will return similar indices each time we cal it.
And names of one table subtables will have neighboring indices.